windows/devices/uwb/UwbDeviceConnector.cxx

```diff
@@ -308,13 +317,152 @@ UwbDeviceConnector::HandleNotifications(wil::shared_hfile handleDriver, std::sto
+        // Invoke callbacks with notification data.
+        // Handle the notification in a fire-and-forget fashion. This may change
+        // later. Since std::async returns a future, and the future's
+        // destructor waits for it to complete, we cannot just ignore the
+        // returned future. To work around this, we move the returned future
+        // into a shared_ptr, then pass this by value to the std::async's
+        // lambda, increasing its reference count. This will ensure the future
+        // is automatically destructed once the async lambda has completed.
+        auto notificationHandlerFuture = std::make_shared<std::future<void>>();
+        *notificationHandlerFuture = std::async(std::launch::async, [this, notificationHandlerFuture, uwbNotificationData = std::move(uwbNotificationData)]() {
+            DispatchCallbacks(std::move(uwbNotificationData));
+        });
+    }
+}
+
+void
+UwbDeviceConnector::OnSessionMulticastListStatus(::uwb::protocol::fira::UwbSessionUpdateMulicastListStatus statusMulticastList)
+{
+    uint32_t sessionId = statusMulticastList.SessionId;
+    auto it = m_sessionEventCallbacks.find(sessionId);
+    if (it == std::end(m_sessionEventCallbacks)) {
+        PLOG_WARNING << "Ignoring MulticastListStatus event due to missing session callback";
+        return;
+    }
+
+    auto& [_, callbacksWeak] = *it;
+    auto callbacks = callbacksWeak.lock();
+    if (not(callbacks->OnSessionMembershipChanged)) {
+        PLOG_WARNING << "Ignoring MulticastListStatus event due to missing session callback";
+        m_sessionEventCallbacks.erase(it);
+        return;
     }
+
+    std::vector<::uwb::UwbPeer> peersAdded;
+    for (const auto& peer : statusMulticastList.Status) {
+        if (peer.Status == UwbStatusMulticast::OkUpdate) {
+            peersAdded.push_back(::uwb::UwbPeer{ peer.ControleeMacAddress });
+        }
+    }
+
+    PLOG_VERBOSE << "Session with id " << statusMulticastList.SessionId << " executing callback for adding peers";
+    callbacks->OnSessionMembershipChanged(peersAdded, {});
+
+    // Now log the bad status
+    IF_PLOG(plog::verbose)
+    {
+        for (const auto& peer : statusMulticastList.Status) {
+            if (peer.Status != UwbStatusMulticast::OkUpdate) {
+                PLOG_VERBOSE << "peer has bad status: " << peer.ToString();
+            }
+        }
+    }
+}
+
+void
+UwbDeviceConnector::OnSessionRangingData(::uwb::protocol::fira::UwbRangingData rangingData)
+{
+    uint32_t sessionId = rangingData.SessionId;
+    auto it = m_sessionEventCallbacks.find(sessionId);
+    if (it == std::end(m_sessionEventCallbacks)) {
+        PLOG_WARNING << "Ignoring RangingData event due to missing session callback";
+        return;
+    }
+    auto callbacks = it->second.lock();
+    if (not(callbacks->OnPeerPropertiesChanged)) {
+        PLOG_WARNING << "Ignoring RangingData event due to missing session callback";
+        return;
+    }

```

Comment:
Can you replicate the approach used in `OnSessionMulticastListStatus` here? That used a structured binding for the map callbacks and erased the callbacks from the map when `lock()` returned a `nullptr`.

Reply:
Thanks, I forgot about this one.